daiquiri/__init__.py

```diff
@@ -51,21 +51,66 @@ def process(self, msg, kwargs):
         return msg, kwargs
 
 
+PARAMETER_TEMPLATE = ' [{0}: {1}]'
+
+
+class ArbitraryContextAdapter(logging.LoggerAdapter):
+    """Logger adapter to add context string to log record's extra data
+
+    Keywords passed to the log call are formatted into a context
+    string, which is then added to the "extra" dictionary passed to
+    the underlying logger so they are emitted with the log message and
+    available to the format string.
+
+    Example:
+      A formatter with the format string of "%(context)s %(messages)"
+      called like logger.info("A message.", test1="a", test2="b")
+      would log the message: " [test1: a] [test2: b] A message."
+
+    Special keywords:
+
+    extra
+      An existing dictionary of extra values to be passed to the
+      logger. If present, the dictionary is copied and extended.
+    """
+
+    def process(self, msg, kwargs):
+        # Make a new extra dictionary combining the values we were
+        # given when we were constructed and a 'context' key which
+        # will contain a formatted string containing anything from
+        # kwargs.
+        extra = self.extra.copy()
+        if 'extra' in kwargs:
+            extra.update(kwargs.pop('extra'))
+        # Format any unknown keyword arguments into the context string.
+        context = ''
+        for name in list(kwargs.keys()):
+            if name == 'exc_info':
+                continue
+            context += PARAMETER_TEMPLATE.format(name, kwargs.pop(name))
+        extra['context'] = context
+        extra['_daiquiri_extra'] = extra
+        kwargs['extra'] = extra
+        return msg, kwargs
```

Comment:
I think you're oversimplifying this problem. Specifically when you say:
> That does not seem too hard to poke into the format string to see what kind of %(foobar) are present and deduce what needs to be in `context.

That is hard. The reason this is hard is because `logging.Formatter` supports all three methods of python formatting (the `%` operator, `"{}".format()`, and string templates). However even only supporting the `%` style formatting functionality in that way would be a hard problem.

Even if I could even find the python source code where `%` is defined (which I can't) to try and match it, that probably would be very difficult since the functionality is described as "similar to using the sprintf() in the C language". `sprintf` basically just calls `vsprintf`, which is defined [here](http://mirror.fsf.org/pmon2000/3.x/src/lib/libc/vsprintf.c). That function is like 100 lines of C, trying to re-implement it in a way that allows me to extract what it would recognize is just asking to introduce so many potential bugs. I could try to create a naive solution but that feels like quite a hack.

Anyway, I can't find an elegant solution to supporting this functionality without breaking backwards compatibility.